sdk/iot/core/src/az_iot_hub_client.c

```diff
@@ -1,12 +1,138 @@
+            + (az_span_length(*module_id) > 0 ? sizeof(hub_client_forward_slash) : 0)
+            + az_span_length(*user_agent)
+            + (az_span_length(*user_agent) > 0 ? sizeof(hub_client_param_separator) : 0)
+            + az_span_length(hub_client_api_version) + sizeof(hub_client_null_terminate)))
+  {
+    return AZ_ERROR_INSUFFICIENT_SPAN_CAPACITY;
+  }
+
+  AZ_RETURN_IF_FAILED(
+      az_span_append(mqtt_user_name, client->_internal.iot_hub_hostname, out_mqtt_user_name));
+  AZ_RETURN_IF_FAILED(
+      az_span_append_uint8(*out_mqtt_user_name, hub_client_forward_slash, out_mqtt_user_name));
+  AZ_RETURN_IF_FAILED(
+      az_span_append(*out_mqtt_user_name, client->_internal.device_id, out_mqtt_user_name));
+  AZ_RETURN_IF_FAILED(
+      az_span_append_uint8(*out_mqtt_user_name, hub_client_forward_slash, out_mqtt_user_name));
+
+  if (az_span_length(*module_id) > 0)
+  {
+    AZ_RETURN_IF_FAILED(az_span_append(*out_mqtt_user_name, *module_id, out_mqtt_user_name));
+    AZ_RETURN_IF_FAILED(
+        az_span_append_uint8(*out_mqtt_user_name, hub_client_forward_slash, out_mqtt_user_name));
+  }
+
+  AZ_RETURN_IF_FAILED(
+      az_span_append(*out_mqtt_user_name, hub_client_api_version, out_mqtt_user_name));
+
+  if (az_span_length(*user_agent) > 0)
+  {
+    AZ_RETURN_IF_FAILED(
+        az_span_append_uint8(*out_mqtt_user_name, hub_client_param_separator, out_mqtt_user_name));
+    AZ_RETURN_IF_FAILED(az_span_append(*out_mqtt_user_name, *user_agent, out_mqtt_user_name));
+  }
+
+  AZ_RETURN_IF_FAILED(
+      az_span_append_uint8(*out_mqtt_user_name, hub_client_null_terminate, out_mqtt_user_name));
+
+  return AZ_OK;
+}
+
+//[Format without module id] {device_id}
+//[Format with module id] {device_id}/{module_id}
+AZ_NODISCARD az_result az_iot_hub_client_id_get(
+    az_iot_hub_client const* client,
+    az_span mqtt_client_id,
+    az_span* out_mqtt_client_id)
+{
+  AZ_PRECONDITION_NOT_NULL(client);
+  AZ_PRECONDITION_VALID_SPAN(mqtt_client_id, 0, false);
+  AZ_PRECONDITION_NOT_NULL(out_mqtt_client_id);
+
+  const az_span* const module_id = &(client->_internal.options.module_id);
+
+  if (az_span_capacity(mqtt_client_id)
+      < (int32_t)(
+            az_span_length(client->_internal.device_id) + az_span_length(*module_id)
+            + (az_span_length(*module_id) > 0 ? sizeof(hub_client_forward_slash) : 0)
+            + sizeof(hub_client_null_terminate)))
+  {
+    return AZ_ERROR_INSUFFICIENT_SPAN_CAPACITY;
+  }
+
+  AZ_RETURN_IF_FAILED(
+      az_span_append(mqtt_client_id, client->_internal.device_id, out_mqtt_client_id));
+
+  if (az_span_length(*module_id) > 0)
+  {
+    AZ_RETURN_IF_FAILED(
+        az_span_append_uint8(*out_mqtt_client_id, hub_client_forward_slash, out_mqtt_client_id));
+    AZ_RETURN_IF_FAILED(az_span_append(*out_mqtt_client_id, *module_id, out_mqtt_client_id));
+  }
+
+  AZ_RETURN_IF_FAILED(
+      az_span_append_uint8(*out_mqtt_client_id, hub_client_null_terminate, out_mqtt_client_id));
```

Comment:
Can we only leak out variable only when we are almost done with function? Something like this?

AZ_RETURN_IF_FAILED(
      az_span_append(mqtt_client_id, client->_internal.device_id, mqtt_client_id));

  if (az_span_length(*module_id) > 0)
  {
    AZ_RETURN_IF_FAILED(
        az_span_append_uint8(mqtt_client_id, hub_client_forward_slash, &mqtt_client_id));
    AZ_RETURN_IF_FAILED(az_span_append(mqtt_client_id *module_id, &mqtt_client_id));
  }

  AZ_RETURN_IF_FAILED(
      az_span_append_uint8(mqtt_client_id hub_client_null_terminate, &mqtt_client_id));

if (out_mqtt_client_id != NULL)
{
   memcpy((void*)out_mqtt_client_id, &mqtt_client_id, sizeof(az_span));
}

Reply:
pushed a change to update the output span only on success. however the output span will have a precondition to not be `NULL`